mlir/lib/ExecutionEngine/conv-validation-wrappers.cpp

```diff
@@ -842,3 +843,192 @@ mcpuConv2dInt8(int64_t rank1, void *f_ptr, int64_t rank2, void *i_ptr,
 }
+
+int findIdxHistRelDiff(double relDiff) {
+  /*
+    hist_relDiff[]:
+    0: 0
+    1: 0 - 1e-6
+    2: 1e-6 - 1e-5
+    3: 1e-5 - 1e-4
+    4: 1e-4 - 1e-3
+    5: 1e-3 - 1e-2
+    6: 1e-2 - 0.1
+    7: 0.1 - 1
+    8: >= 1
+   */
+  if (relDiff == 0.0)
+    return 0;
+  if (relDiff < 1.0e-06)
+    return 1;
+  if (relDiff < 1.0e-05)
+    return 2;
+  if (relDiff < 1.0e-04)
+    return 3;
+  if (relDiff < 1.0e-03)
+    return 4;
+  if (relDiff < 1.0e-02)
+    return 5;
+  if (relDiff < 0.1)
+    return 6;
+  if (relDiff < 1.0)
+    return 7;
+  return 8;
+}
+
+template <typename T>
+void mcpuVerify(T *gpuResults, T *validationResults, int64_t dataSize,
+                float thr_RMS, float thr_absDiff, float thr_relDiff) {
+  float valNum, gpuNum;
+  // metric maxAbsDiff
+  float maxAbsDiff = 0.0f;
+  double sumAbsDiff = 0.0;
+  float maxVAL_abs = 0.0f;
+  float maxGPU_abs = 0.0f;
+  // metric maxRelDiff
+  double maxRelDiff = 0.0;
+  double sumRelDiff = 0.0;
+  float maxVAL_rel = 0.0f;
+  float maxGPU_rel = 0.0f;
+  // Metric RMS
+  float maxMag = 0.0f;
+  double sumDiffSq = 0.0;
+  // histogram
+  // hist_relDiff[9]++ when cpuVal == 0
+  int hist_relDiff[10] = {0};
+  for (int64_t i = 0; i < dataSize; ++i) {
+    valNum = static_cast<float>(validationResults[i]);
+    gpuNum = static_cast<float>(gpuResults[i]);
+    // Update the max magnitutde value
+    float maxNum = std::max(fabs(valNum), fabs(gpuNum));
+    maxMag = std::max(maxMag, maxNum);
+
+    if (valNum == gpuNum) {
+      hist_relDiff[0]++;
+    } else {
+      float absDiff = fabs(valNum - gpuNum);
+      // Update maxAbsDiff and its correspinding pair of values
+      if (absDiff > maxAbsDiff) {
+        maxVAL_abs = valNum;
+        maxGPU_abs = gpuNum;
+        maxAbsDiff = absDiff;
+      }
+      sumAbsDiff += static_cast<double>(absDiff);
+      // Update maxRelDiff only if cpuVal != 0
+      double relDiff = 0.0;
+      if (valNum != 0.0f) {
+        relDiff =
+            static_cast<double>(absDiff) / (static_cast<double>(fabs(valNum)));
+        hist_relDiff[findIdxHistRelDiff(relDiff)]++;
+        if (relDiff > maxRelDiff) {
+          maxVAL_rel = valNum;
+          maxGPU_rel = gpuNum;
+          maxRelDiff = relDiff;
+        }
+        sumRelDiff += relDiff;
+      } else {
+        hist_relDiff[9]++;
+      }
+      // Accumulate square root
+      sumDiffSq += static_cast<double>(absDiff) * static_cast<double>(absDiff);
+      // Print out values if difference is larger than threshold
+      if (absDiff > thr_absDiff || relDiff > thr_relDiff)
+        printf("%ld: %f %f %f %lf\n", i, valNum, gpuNum, absDiff, relDiff);
+    }
+  }
+  double aveAbsDiff = sumAbsDiff / static_cast<double>(dataSize);
+  double aveRelDiff = sumRelDiff / static_cast<double>(dataSize);
+  double err_RMS = sqrt(sumDiffSq) / (static_cast<double>(maxMag) *
+                                      sqrt(static_cast<double>(dataSize)));
+  // Verbose information about the difference
```

Comment:
For the switching point, would it be too much details for `miopen-gen` users?
When `check-mlir-miopen` is invoked, `lit` only matches the last line of the output, i.e. `[1 1 1]`. 
When `miopen-gen` is invoked explicitly, I think the developer expects to see those debug information. Therefore, the on/off functionality is fulfilled by the way `miopen-gen` is used.

I'll refactor the histogram printing process.
> BTW, why not use the more modern cout for printing?

I do not see any reasons not using it. I can switch to it.

Reply:
Here are my thoughts on the use case when we want to turn print debug messages to off:
 - In automated testing, if we'd only match [1 1 1] then we should try to limit the error printing to only those. From robustness point of view.
 - In development setting, it could be tedious to always see a bunch of information printed on screen for a passing test.

Reply:
I agree on the development case.

However, I'll note that `FileCheck` suppresses output of passing tests.

If we're worried about false-positive matches, we can add some sort of prefix to the result line - say `RESULT: `, to make it harder to accidentally write a debug print that makes the test pass

Reply:
Also, @jerryyin , I'd actually prefer `printf()` here. It makes it much easier to control things like float precision. `cout`'s API for that is ugly